### PR TITLE
Wallet recovery

### PR DESCRIPTION
**Pull request by blakecduncan**
Merging is blocked by https://github.com/web3well/bls-wallet/pull/367

## What is this PR doing?

- Adding an addRecoveryWallet method to the RPC
- UI changes to allow the user to recover

Video demo of recovery
https://www.loom.com/share/34e84b23cd8647058dc92afc43a9ed28

## How can these changes be manually tested?

- You can test by recovering from an instant wallet to the Quill extension (as seen in the demo video above)
- This won't currently work on Goerli because there is a bug fix in progress.  If you want to test it locally you need to use `bls-wallet-clients@0.8.0-2d28b7b`

## Does this PR resolve or contribute to any issues?

https://github.com/web3well/bls-wallet/issues/281

## Checklist
- [x] I have manually tested these changes
- [x] Post a link to the PR in the group chat

## Guidelines
- If your PR is not ready, mark it as a draft
- The `resolve conversation` button is for reviewers, not authors
  - (But add a 'done' comment or similar)

### CODE REVIEW

File: extension/source/background/KeyringController.ts
Comment (blakecduncan) [inline extension/source/background/KeyringController.ts]: I think calling this function brings up the transaction approval modal.  I'm not sure if we want that to happen or if I should just call the method directly doing something like

```
  const bundle = signerWallet.sign({
    nonce: await signerWallet.Nonce(),
    actions: [
      {
        ethValue: 0,
        contractAddress: verificationGatewayContract.address,
        encodedFunction:
          verificationGatewayContract.interface.encodeFunctionData(
            'recoverWallet',
            [
              addressSignature,
              instantWalletHash,
              recoverySaltHash,
              newWallet.PublicKey(),
            ],
          ),
      },
    ],
  });
```

File: extension/source/background/KeyringController.ts
Comment (blakecduncan) [inline extension/source/background/KeyringController.ts]: This is related to the above comment.  I need to test this more but I might need to figure out a way to get this to wait until the transaction is approved by the user before calling the addAccount function.

File: extension/source/background/KeyringController.ts
Comment (blakecduncan) [inline extension/source/background/KeyringController.ts:189]: @kautukkundan There's one thing I wanted to raise that I'm doing differently than what we discussed on our call today.

I'm now accepting a `signerWalletPrivatekey` as a param here.  That's because, based off our convo earlier, you are going to generate a wallet on the fly to be used for the recovery.  This way, you can just pass that wallet's PK to this function.

File: extension/source/background/KeyringController.ts
Comment (blakecduncan) [inline extension/source/background/KeyringController.ts:342]: I also didn't change this to use the `eth_sendTransaction` function because that function uses the selected wallet in the quill extension.  If we are generating a wallet on the fly, I think we will need to create the bundle ourselves here.  Fyi @kautukkundan

File: extension/source/Home/Wallet/Wallets/Recovery/RecoverWalletModal.tsx
Comment (JohnGuilding) [inline extension/source/Home/Wallet/Wallets/Recovery/RecoverWalletModal.tsx:75]: I accidentally clicked next on the second page without copying the address and found that I couldn't go back. I instead had to close to model or refresh the page (can't remember which) to start the process again and access the address.

Wondering if there is an easy way to go back a step?

File: extension/source/Home/Wallet/Wallets/Recovery/RecoverWalletModal.tsx
Comment (JohnGuilding) [inline extension/source/Home/Wallet/Wallets/Recovery/RecoverWalletModal.tsx:37]: Do we need to explicitly add the type here? Can see that we're letting typescript infer the type for the above useState's

File: extension/source/Home/Wallet/Wallets/Recovery/StepOneInfo.tsx
Comment (JohnGuilding) [inline extension/source/Home/Wallet/Wallets/Recovery/StepOneInfo.tsx]: Thoughts on a slightly different wording here instead of 'stuff'? Maybe something like:

1. "...requires you to copy-paste two values from..." or 
2. "...requires you to copy-paste two pieces of information from..."

File: extension/source/Home/Wallet/Wallets/Recovery/StepOneInfo.tsx
Comment (JohnGuilding) [inline extension/source/Home/Wallet/Wallets/Recovery/StepOneInfo.tsx]: Thoughts on:

"...steps that follow, else you..."

File: extension/source/Home/Wallet/Wallets/Recovery/StepThreeRecover.tsx
Comment (JohnGuilding) [inline extension/source/Home/Wallet/Wallets/Recovery/StepThreeRecover.tsx:14]: Similar point here and line below, do we want to explicitly state the type?

File: extension/source/Home/Wallet/Wallets/Recovery/StepThreeRecover.tsx
Comment (JohnGuilding) [inline extension/source/Home/Wallet/Wallets/Recovery/StepThreeRecover.tsx:41]: Haven't seen this way of adding classNames before - nice. Might be a silly question but is there a reason why you would pick this way over say the syntax on line 26?

File: extension/source/Home/Wallet/Wallets/Recovery/StepTwoWalletCreation.tsx
Comment (JohnGuilding) [inline extension/source/Home/Wallet/Wallets/Recovery/StepTwoWalletCreation.tsx]: Tiny bit of whitespace :)

File: extension/source/background/KeyringController.ts
Comment (JohnGuilding) [inline extension/source/background/KeyringController.ts]: Nitpicky but just asking for consistency's sake, should the word "the" after `signerWalletPrivateKey` be capitalised here? Or should the capitalised words after the param names above this be uncapitalised?

File: extension/source/background/KeyringController.ts
Comment (JohnGuilding) [inline extension/source/background/KeyringController.ts:314]: Might not be fully understanding here, my two questions are:

1. Is this creating a signature using a randomly generated private key and an already generated contract address? 

2. If so, how is this linking an already created address to a randomly generated private key? My knowledge of ECDSA cryptography is that the address must be derived from the public key, which is derived from the private key. But I admit my knowledge may be lacking here as these are smart contract wallets and we're not using ECDSA? And I don't fully understand how they are instantiated.

File: extension/source/background/KeyringController.ts
Comment (JohnGuilding) [inline extension/source/background/KeyringController.ts:119]: Same question for `addRecoveryWallet()`, do we need to add unit tests for this new logic? I'm assuming as this is the extension which isn't meant to be production code, we are happy with not adding tests?

**Review by JohnGuilding — COMMENTED** [review]
Tested locally and works well. Was able to recover a wallet successfully from the instant wallet into Quill.

I'm not too confident reviewing the changes in the `KeyringController.ts` as it's quite a lot of new stuff for me, so would be good to get a second opinion on that specific file :)

File: extension/source/Home/Wallet/Wallets/Recovery/StepThreeRecover.tsx
Comment (jacque006) [inline extension/source/Home/Wallet/Wallets/Recovery/StepThreeRecover.tsx:22]: Do we need an error state/handling if this RPC call fails?

File: extension/source/Home/Wallet/Wallets/Recovery/StepTwoWalletCreation.tsx
Comment (jacque006) [inline extension/source/Home/Wallet/Wallets/Recovery/StepTwoWalletCreation.tsx]: Why are deps disabled here? Should just be `rpc`.

File: extension/source/Home/Wallet/Wallets/Recovery/StepTwoWalletCreation.tsx
Comment (jacque006) [inline extension/source/Home/Wallet/Wallets/Recovery/StepTwoWalletCreation.tsx:60]: What rule was disabled here? Better to disable a specific rule than all linting rules.

File: extension/source/background/KeyringController.ts
Comment (jacque006) [reply extension/source/background/KeyringController.ts:119]: @JohnGuilding We generally don't have tests for the extension as they would be complicated to setup & manage. However, as extension gets more stable might make sense to add some e2e tests for more common paths. Would save us some regression testing as well.

File: extension/source/background/KeyringController.ts
Comment (jacque006) [inline extension/source/background/KeyringController.ts]: Use a `BN254` spaced key instead of just a random hex. See https://github.com/thehubbleproject/hubble-bls/blob/master/src/mcl.ts#L186 . We should expose that via the signer in the client library somewhere.

File: extension/source/background/KeyringController.ts
Comment (jacque006) [inline extension/source/background/KeyringController.ts]: See above comment on BLS private key gen.

**Review by jacque006 — COMMENTED** [review]
It might also be beneficial to move some of the functionality into the clients lib so that other wallets/dApps can take advantage of this functionality, similar to https://github.com/web3well/bls-wallet/issues/220 . But I would be okay if that's follow up work.

File: extension/source/background/KeyringController.ts
Comment (blakecduncan) [reply extension/source/background/KeyringController.ts]: @jacque006  Ahh interesting we do `generateRandomHex` a couple of other places in the keyring controller which is where I copied this from.  Should I also switch those?

Also, just curious, what is the benefit of `BN254` over a random hex?

Comment (blakecduncan) [reply extension/source/background/KeyringController.ts]: Just want to double check that this is the right way to call that function?

```
randFr().serializeToHexStr()
```

File: extension/source/background/KeyringController.ts
Comment (blakecduncan) [reply extension/source/background/KeyringController.ts:314]: @JohnGuilding good question

1. Yeah it is.  It's the smart contract address of the 'instant wallet'.  So the wallet we are recovering.  Signed by the newly generated private key that will become the new private key for the address we are recovering.

3. I may not be the best at explaining this but this is my understanding. This signature is basically how the trusted wallet (the one that was set in the recovery hash) says that this private key now has permission to act on behalf of the instant wallet address.  When the recovery happens this private key will create a new public key but instead of deterministically generating an address the public key will map to the existing instant wallet address.  This signature is a way of verifying and setting this PK as the PK to the old address without revealing the private key.

File: extension/source/Home/Wallet/Wallets/Recovery/StepThreeRecover.tsx
Comment (kautukkundan) [reply extension/source/Home/Wallet/Wallets/Recovery/StepThreeRecover.tsx:41]: due to tailwind-css the line length for classes sometimes exceeds 80 which makes the linter unhappy.

File: extension/source/Home/Wallet/Wallets/Recovery/RecoverWalletModal.tsx
Comment (kautukkundan) [reply extension/source/Home/Wallet/Wallets/Recovery/RecoverWalletModal.tsx:37]: My personal preference is to have explicit typing everywhere. Copied the above 2 lines from somewhere else in the codebase so missed adding type to it.

File: extension/source/Home/Wallet/Wallets/Recovery/StepTwoWalletCreation.tsx
Comment (kautukkundan) [reply extension/source/Home/Wallet/Wallets/Recovery/StepTwoWalletCreation.tsx:60]: it wanted to also have a "keyboard input" for the copy button which didn't make much sense to me.

File: extension/source/Home/Wallet/Wallets/Recovery/StepTwoWalletCreation.tsx
Comment (kautukkundan) [reply extension/source/Home/Wallet/Wallets/Recovery/StepTwoWalletCreation.tsx]: I sometimes don't understand these deps. Like it's not possible to change some deps without explicit reload which is gonna rerender anyway.

But yeah, let's add rpc here.

File: extension/source/Home/Wallet/Wallets/Recovery/StepThreeRecover.tsx
Comment (kautukkundan) [reply extension/source/Home/Wallet/Wallets/Recovery/StepThreeRecover.tsx:22]: error-handling is not great across the extension's codebase. I think we should have a separate task for improving errors/fails altogether. That's why left it simple here as well.

File: extension/source/Home/Wallet/Wallets/Recovery/RecoverWalletModal.tsx
Comment (JohnGuilding) [reply extension/source/Home/Wallet/Wallets/Recovery/RecoverWalletModal.tsx:37]: I may lean towards letting typescript infer the type but don't have a strong opinion either way, so happy with either option as long as we're being consistent :)

File: extension/source/Home/Wallet/Wallets/Recovery/StepTwoWalletCreation.tsx
Comment (JohnGuilding) [reply extension/source/Home/Wallet/Wallets/Recovery/StepTwoWalletCreation.tsx]: Useful comment on this issue with links for further reading from Dan Abramov
https://github.com/facebook/create-react-app/issues/6880#issuecomment-485912528

File: extension/source/background/KeyringController.ts
Comment (JohnGuilding) [reply extension/source/background/KeyringController.ts:314]: Thanks for the answer, have a better understanding now. Will make sure to re-read your recovery article too

File: extension/source/background/KeyringController.ts
Comment (kautukkundan) [reply extension/source/background/KeyringController.ts]: yeah makes sense 

> what is the benefit of BN254 over a random hex?

If I understand correctly, The random number generated might sometimes lie outside the range of the BN curve we are using. This function ensures it's in range.

Comment (jacque006) [reply extension/source/background/KeyringController.ts]: @kautukkundan correct. If the key is out of range, we just hash it into a key that is in range. I have a bucklist item/change that will be more explicit about this in the future https://github.com/thehubbleproject/hubble-bls/issues/30

File: extension/source/Home/Wallet/Wallets/Recovery/RecoverWalletModal.tsx
Comment (blakecduncan) [reply extension/source/Home/Wallet/Wallets/Recovery/RecoverWalletModal.tsx:75]: Fixed!

File: extension/source/Home/Wallet/Wallets/Recovery/RecoverWalletModal.tsx
Comment (voltrevo) [reply extension/source/Home/Wallet/Wallets/Recovery/RecoverWalletModal.tsx:37]: @kautukkundan Are you sure that's your preference in this case where TypeScript is correctly inferring the type? Would you also add `<boolean>` and `<number>` to the previous lines?

Comment (kautukkundan) [reply extension/source/Home/Wallet/Wallets/Recovery/RecoverWalletModal.tsx:37]: @voltrevo Fine with me as is in this _particular_ case. 

The problem starts happening when using complex types. eg - let's say a new `Address` type is defined as a string that always starts with "0x". In this case initializing with default '' would also work as it'll infer type string, but may break code later when things get complex. This had been the case earlier when using sol.G1 and sol.G2 types in the HubbleBLS library.

So looking at broader picture and anticipating future I prefer being explicit even with simpler types to maintain consistency, If that makes sense.

**Review by kautukkundan — APPROVED** [review]
LGTM 🚀

**Comment by Nerotool** [comment]
SEND A DM TO NEROTOOL_ ON INSTAGRAM FOR HACKING

**Pull request merged**
